### PR TITLE
[REVIEW] Concatenation of different typed columns

### LINKED ISSUE #1881: [FEA] Concatenation of dataframes with columns of different dtypes
Users may want to concatenate dataframes/series of differing types, which is currently blocked by

https://github.com/rapidsai/cudf/blob/8f7fe88fefd3ba3a6d081b4a226eee5b5d97c175/python/cudf/dataframe/column.py#L91-L93

One potential approach to this problem is to upcast types where necessary. For example:

```python
import pandas as pd
​
pdf1 = pd.DataFrame({'a':np.array([1,2,3], dtype=np.float32),
                     'b':np.array([1,2,3], dtype=np.float32),
                     'c':np.array([1,2,3], dtype=np.int16),
                     'd': ['a','b','c']
             })
​
pdf2 = pd.DataFrame({'a':np.array([1,2,3], dtype=np.float64),
                     'b':np.array([1,2,3], dtype=np.float64),
                     'c':np.array([1,2,3], dtype=np.int64),
                     'd': ['a','b','c']
             })
​
pd.concat([pdf1, pdf2]).dtypes
a    float64
b    float64
c      int64
d     object
dtype: object
```

This would resolve rapidsai/dask-cudf#264, though that my be resolved independently in Dask.

This is related to but different than #188

### PR DESCRIPTION
**Pull request by madsbk**
Fixes #1881 by implementing concatenation of dataframes with columns of different dtypes.

It uses [`np.find_common_type()`](https://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.find_common_type.html) to find a suitable data type and then cast the columns before the concatenation.

### CODE REVIEW

File: python/cudf/dataframe/column.py [hunk 1]
```diff
@@ -62,6 +62,14 @@ def _concat(cls, objs, dtype=None):
                 dtype = np.dtype(dtype)
                 return Column(Buffer.null(dtype))
 
+        # Find the dtype of each column ignoring pure NULL columns
+        col_dtypes = [o.dtype for o in filter(lambda o: len(o) != o.null_count, objs)]
+        # Use NumPy to find a common dtype
+        common_dtype = np.find_common_type(col_dtypes, [])
```
Comment (kkraus14) [inline python/cudf/dataframe/column.py]: This won't handle category dtypes, so you'd need to check for the category handling before doing this.

Comment (madsbk) [reply python/cudf/dataframe/column.py]: Ahh I see, is there coercion rules between categories? 
Or should we raise an error when concatenating columns of different category types?

Comment (kkraus14) [reply python/cudf/dataframe/column.py]: I believe we already have logic to synchronize the categories before concatenation, the main thing is we should guard against trying to concatenate a category column or a string column with numeric dtype columns.

Comment (kkraus14) [reply python/cudf/dataframe/column.py]: And in this comment I specifically meant numpy won't know what to do with `"category"` or a `CategoricalDtype` object

Comment (madsbk) [reply python/cudf/dataframe/column.py]: I see, so it should only try to do casting if all columns are of regular NumPy dtypes?

**Comment by madsbk** [comment]
rerun test

**Comment by madsbk** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

File: python/cudf/tests/utils.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
-
 import numpy as np
 import pandas as pd
 from cudf.utils import utils
 
 import pandas.util.testing as tm
 
+supported_numpy_dtypes = ['bool', 'int8', 'int16', 'int32',
+                          'int64', 'float32', 'float64']
```
Comment (kkraus14) [inline python/cudf/tests/utils.py]: We should handle `datetime64[ms]` here as well.

**Comment by madsbk** [comment]
@kkraus14, I have added `datetime64[ms]` and `datetime64[us]` to `supported_numpy_dtypes`. However, concatenating `datetime` with regular dtypes such as `int64` produces columns of `object` in pandas. Currently in this case, cudf will raise `ValueError`. Is that alright?

**Comment by kkraus14** [comment]
@madsbk just checking in here, are you swamped elsewhere and someone else should pick this up?

**Comment by BartleyR** [comment]
This is something we run into a lot for our use cases (1 CSV out of 1000s has one value that is interpreted as a different datatype), so this is something we'd use a lot.

**Comment by madsbk** [comment]
Sorry, I got sidetracked by other stuff. I will make sure this will get merged.

**Comment by madsbk** [comment]
@kkraus14, I think this PR is ready for merging.

File: python/cudf/cudf/dataframe/column.py [hunk 3]
```diff
@@ -61,6 +62,28 @@ def _concat(cls, objs, dtype=None):
...
+        # we cast them to a common dtype.
+        # Notice, we can always cast pure null columns
+        not_null_cols = list(filter(lambda o: len(o) != o.null_count, objs))
+        if len(not_null_cols) > 0 and (
+            len(
+                [
+                    o
+                    for o in not_null_cols
+                    if not isinstance(o, NumericalColumn)
+                    or np.issubdtype(o.dtype, np.datetime64)
```
Comment (kkraus14) [inline python/cudf/cudf/dataframe/column.py]: Is this to prevent trying to concatenate say an `int16` with a `datetime64[s]`?

Comment (madsbk) [reply python/cudf/cudf/dataframe/column.py]: Yes, in this case I think it makes sense to raise an exception

**Pull request merged**